### PR TITLE
Replace `EffElem` with `Data.List.Elem`

### PR DESCRIPTION
**Pull request by ericqweinstein**
Proposed fix for https://github.com/idris-lang/Idris-dev/issues/3068. @jfdm, is this what you were thinking?

### CODE REVIEW

**Comment by ericqweinstein** [comment]
Taking a look later today:

```
Effects.idr:194:52:When checking type of Effects.updateResTy:
Type mismatch between
        Type (Type of Elem _ _)
and
        _ -> _ (Is Elem _ _ applied to too many arguments?)
```

**Comment by ericqweinstein** [comment]
@jfdm Were you looking to replace `EffElem` with `Data.List.Elem` entirely, or redefine the former in terms of the latter? (I guess I'm a bit confused because `EffElem : Effect -> Type -> List EFFECT -> Type`, whereas `Data.List.Elem : a -> List a -> Type`.)

**Comment by ahmadsalim** [comment]
@ericqweinstein I believe it should generally be something in the form of `Elem (MkEff a e) xs`, because that is what the `Here` constructor in `EffElem` checked for. That is maybe why you are currently experiencing issues with compiling things.

**Comment by ericqweinstein** [comment]
@ahmadsalim That makes perfect sense—thanks!

File: libs/effects/Effects.idr [hunk 1]
```diff
@@ -346,7 +340,7 @@ new = New
 -- ---------------------------------------------------------- [ an interpreter ]
 
 private
-execEff : Env m xs -> (p : EffElem e res xs) ->
+execEff : Env m xs -> (p : Data.List.Elem (MkEff a e) xs) ->
```
Comment (ahmadsalim) [inline libs/effects/Effects.idr]: This should be `MkEff res e`, which is why the test fails.

Comment (ericqweinstein) [inline libs/effects/Effects.idr]: My mistake, thanks for catching! Should have this ready later today.

**Comment by ericqweinstein** [comment]
I'm now seeing:

```
./Effect/Exception.idr:32:7:
When checking right hand side of raise with expected type
        EffM m b [EXCEPTION a] (\v => [EXCEPTION a])
No such variable Data.List.Elem
```

Which suggests to me a failure to type check (as `Data.List.Elem` [does exist](https://github.com/idris-lang/Idris-dev/blob/master/libs/base/Data/List.idr#L8)), but I'm not totally sure where I'm afield here. (I'm very new to Idris, so apologies for taking so long with this one.)

**Comment by ahmadsalim** [comment]
@ericqweinstein Hmm, I am unsure why this happens.
Is there any reason why you don't just use `Elem`?

PS: It is totally fine that you take some time on this :). Although, please do compile your code locally before pushing to catch errors quicker.

**Comment by ahmadsalim** [comment]
@ericqweinstein So apparently, it seems that you need to add `import Data.List` in the top of `Effect.Exception` for this to work (and all other files that uses the particular public export function).

**Comment by jfdm** [comment]
You are best to set the import to:

```
import public Data.List
```

This will save you having to add it to all the other effects files.

**Comment by ericqweinstein** [comment]
@ahmadsalim @jfdm Thanks! I'll make the import public and make sure things look good after compiling locally before pushing up again. As for the full qualification, I'd wanted to avoid conflicts with `Data.Vect.Elem` (see [here](https://github.com/idris-lang/Idris-dev/blob/master/libs/base/Data/Vect.idr#L578)), but if there's a preferred way of specifying `Data.List.Elem` without repeating myself each time, I'm happy to do that.

**Comment by ericqweinstein** [comment]
Okay, `effects/` type checks and compiles. Pushing up!

**Comment by ahmadsalim** [comment]
@ericqweinstein Thanks for the contribution!

**Comment by ahmadsalim** [comment]
@ericqweinstein One small request though, would it be possible to investigate whether one could use `Elem` unqualified instead of `Data.List.Elem` everywhere?

I think the code would like nicer without qualification.

**Comment by ericqweinstein** [comment]
@ahmadsalim Of course! I'll do that today.

**Comment by ericqweinstein** [comment]
@ahmadsalim What do you think of `with List Elem` as a solution here?

**Comment by ahmadsalim** [comment]
@ericqweinstein If plain `Elem` does not work then what you have now is fine.

**Comment by ahmadsalim** [comment]
Fixes #3068

**Comment by ericqweinstein** [comment]
@ahmadsalim Yeah, there was a conflict between `Data.Vect.Elem` and `Data.List.Elem`. Thanks for merging!

**Comment by ahmadsalim** [comment]
@ericqweinstein Thank you!

**Pull request merged**